ui-core/src/components/inputs/InputWithSuggestions.tsx

```diff
@@ -0,0 +1,107 @@
+import React, {
+  useState,
+  useRef,
+  useEffect,
+  ChangeEventHandler,
+  KeyboardEventHandler,
+  ReactNode,
+} from 'react';
+import cx from 'classnames';
+import Input, { InputProps } from './Input';
+
+interface InputWithSuggestionsProps<T> extends Omit<InputProps, 'onChange' | 'value'> {
+  suggestions: Array<T>;
+  onChange: (value: T) => void;
+  getSuggestionLabel: (option: T) => string;
+  getSuggestionKey: (value: T) => string;
+  customLabel?: ReactNode;
+}
+
+function InputWithSuggestions<T>({
+  suggestions,
+  onChange,
+  getSuggestionLabel,
+  getSuggestionKey,
+  customLabel,
+  ...inputProps
+}: InputWithSuggestionsProps<T>) {
+  const [filteredSuggestions, setFilteredSuggestions] = useState<T[]>([]);
+  const [activeSuggestion, setActiveSuggestion] = useState<number>(-1);
+  const [showSuggestions, setShowSuggestions] = useState<boolean>(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (e) => {
+    const userInput = e.currentTarget.value;
+
+    if (userInput === '') {
+      setFilteredSuggestions([]);
+      setShowSuggestions(false);
+      return;
+    }
+
+    const filtered = suggestions.filter((suggestion) =>
+      getSuggestionLabel(suggestion).toLowerCase().startsWith(userInput.toLowerCase())
+    );
+
+    setFilteredSuggestions(filtered);
+    setShowSuggestions(true);
+  };
+
+  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (e) => {
+    if (e.key === 'ArrowDown') {
+      setActiveSuggestion((prev) => (prev < filteredSuggestions.length - 1 ? prev + 1 : prev));
+    } else if (e.key === 'ArrowUp') {
+      setActiveSuggestion((prev) => (prev > 0 ? prev - 1 : prev));
+    } else if ((e.key === 'Enter' || e.key === 'Tab') && activeSuggestion >= 0) {
+      selectSuggestion(activeSuggestion);
+    } else if (e.key === 'Escape') {
+      setShowSuggestions(false);
+    }
```

Comment:
Doesn't change the behavior but seems more logical ?

```suggestion
    } else if (e.key === 'Escape') {
      setShowSuggestions(false);
      setFilteredSuggestions([]);
    }
```